Update subscribe and assign method to follow same processing logic as apache kafka

haijun2022 opened the pull request:
This PR fixed the bug described in issue https://github.com/strimzi/strimzi-kafka-bridge/issues/756

it update subscribe and assign method to follow same processing logic as apache kafka.

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/KafkaBridgeConsumer.java:
```diff
@@ -93,8 +93,12 @@ public void create(Properties config, String groupId) {
+
         if (topicSubscriptions.isEmpty()) {
-            throw new IllegalArgumentException("At least one topic to subscribe has to be specified!");
+            this.unsubscribe();
```
so after unsubscribing you are actually not returning and the subscribe operation will run anyway. It should not cause a problem just because the Kafka subscribe with empty topic list will behave as an unsubscribe.
For this reason I would say that this addition about calling unsubscribe is not needed, we should just allow an empty topics list and then allowing the code to run the subscribe.

ppatierno commented on src/main/java/io/strimzi/kafka/bridge/KafkaBridgeConsumer.java:
```diff
@@ -151,6 +155,10 @@ public void assign(List<SinkTopicSubscription> topicSubscriptions) {
 
+        if (topicPartitions.isEmpty()) {
+            this.unsubscribe();
+        }
```
same as described before. Run the unsubscribe here, not returning and then the code runs the assign.
Just running the assign with an empty topics list should work.

haijun2022 replied:
Thanks for the review, exactly, will update the code!

haijun2022 commented:
@ppatierno @scholzj I have updated the code and added test code based on your suggestions. Could you reivew it again and let me know if there is any suggestions. thank you!

ppatierno commented on src/test/java/io/strimzi/kafka/bridge/http/ConsumerSubscriptionIT.java:
```diff
@@ -160,6 +161,62 @@ void subscriptionConsumerDoesNotExistBecauseNotCreated(VertxTestContext context)
+        consumerJson.put("format", "json");
+        consumerService()
+                .createConsumer(context, groupId, consumerJson)
+                .subscribeConsumer(context, groupId, name, topicsRoot);
```
after subscribing you should verify that it was ok by listing the subscription, so that after unsubscribe you can do the same and checking it's empty.
Anyway, for how the Kafka client works, after the subscribe operation you need to execute some poll/consume in order to allow the consumer to be really subscribed (joining the consumer group). The subscribe operation itself just fill a list inside the Kafka consumer nothing more.
See the `listConsumerSubscriptions` test which does something like that after subscribe and before listing the subscriptions.

ppatierno commented on src/test/java/io/strimzi/kafka/bridge/http/ConsumerSubscriptionIT.java:
```diff
@@ -355,6 +412,60 @@ void assignAfterSubscriptionTest(VertxTestContext context) throws InterruptedExc
+        JsonObject partitionsRoot = new JsonObject();
+        JsonArray partitions = new JsonArray();
+
+        partitionsRoot.put("partitions", partitions);
```
compared to the previous test, there is no assignment to a partition and then assignment with empty list to validate the feature.

haijun2022 replied:
just added the section, thank you!

ppatierno commented:
@haijun2022 I see you are on the right way! I noticed you added the check I suggested on the subscription test only but it should be on the assign test as well.

haijun2022 commented:
> @haijun2022 I see you are on the right way! I noticed you added the check I suggested on the subscription test only but it should be on the assign test as well.

Thank you for the review, I tried but did not find a way to go to that logic branch as you suggested so I removed that test code! Will double check, meanwhile please let me know if that is doable!

ppatierno commented:
> Thank you for the review, I tried but did not find a way to go to that logic branch as you suggested so I removed that test code! Will double check, meanwhile please let me know if that is doable!

I am not sure what you mean by this :-(
You added the logic to unsubscribe when someone use `assign` but with empty partitions.
You added a test for it. But it was missing the suggestions I made to you on the other test (the one about unsubscribe via a `subscribe` with empty topics collection).

ppatierno commented:
@haijun2022 do you have any idea when this PR will be ready? We are in the process to start a 0.25.0 release and having your contribution included would be great :-)

haijun2022 commented:
Hi @ppatierno, I am working on it now and will update by the end of today, thank you!

haijun2022 commented:
> Hi @ppatierno, I am working on it now and will update by the end of today, thank you!

Hi @ppatierno , I just updated the code and please help review on it.  My IDE added some whitespaces which is not my intension, please ignore them during your review. Sorry for that!

ppatierno commented:
@haijun2022 it would be great you remove those whitespaces. They are noise in the PR and we don't want to merge the PR with them, thanks!

haijun2022 commented:
> 

Yes, let me know if others are good now? I will update them later

ppatierno commented on src/test/java/io/strimzi/kafka/bridge/http/ConsumerSubscriptionIT.java:
```diff
@@ -160,6 +161,92 @@ void subscriptionConsumerDoesNotExistBecauseNotCreated(VertxTestContext context)
+                });
+        consume.get(TEST_TIMEOUT, TimeUnit.SECONDS);
+
+        CompletableFuture<Boolean> subscribe = new CompletableFuture<>();
```
can you create this later when it is really used just before the subscription to an empty topics list? Maybe even calling `subscribeEmpty` would be better. Thanks!

haijun2022 replied:
updated, thanks!

ppatierno commented on src/test/java/io/strimzi/kafka/bridge/http/ConsumerSubscriptionIT.java:
```diff
@@ -160,6 +161,92 @@ void subscriptionConsumerDoesNotExistBecauseNotCreated(VertxTestContext context)
+
+        subscribe.get(TEST_TIMEOUT, TimeUnit.SECONDS);
+        // Validate the existing consumer list
+        CompletableFuture<Boolean> listSubscriptionsAfter = new CompletableFuture<>();
```
you can reuse the `listSubscriptions` variable as before just creating a `new CompletableFuture<>();`

ppatierno commented on src/test/java/io/strimzi/kafka/bridge/http/ConsumerSubscriptionIT.java:
```diff
@@ -355,6 +442,105 @@ void assignAfterSubscriptionTest(VertxTestContext context) throws InterruptedExc
+        assignEmptyCF.get(TEST_TIMEOUT, TimeUnit.SECONDS);
+
+        // Validate the existing consumer list
+        CompletableFuture<Boolean> listSubscriptionsAfter = new CompletableFuture<>();
```
same as before you should be able to use just one `listSubscriptions` completable future for before and after, just creating a new instance assigning to the same variable.

haijun2022 replied:
Hi @ppatierno , that does not work here. it shows "Local variable listSubscriptionsBefore defined in an enclosing scope must be final or effectively finalJava(536871575)"

ppatierno commented:
@haijun2022 I think these are last comments and we are good to go.

ppatierno reviewed (approved):
LGTM. Thanks a lot for the contribution!

haijun2022 commented:
> LGTM. Thanks a lot for the contribution!

Thanks a lot for your reivew and suggestions. @ppatierno
